src/shadowbox/server/server_access_key.ts

```diff
@@ -90,73 +123,162 @@ export class ServerAccessKeyRepository implements AccessKeyRepository {
     this.keyConfig.data().nextId += 1;
     const metricsId = uuidv4();
     const password = generatePassword();
-    // Save key
-    const accessKeyJson: AccessKeyConfig = {
+    const accessKey: AccessKey = {
       id,
-      metricsId,
       name: '',
-      port,
-      encryptionMethod: this.NEW_USER_ENCRYPTION_METHOD,
-      password
+      metricsId,
+      proxyParams: {
+        hostname: this.proxyHostname,
+        portNumber: port,
+        encryptionMethod: this.NEW_USER_ENCRYPTION_METHOD,
+        password,
+      },
+      quota: undefined,
+      isOverQuota: false
     };
-    this.keyConfig.data().accessKeys.push(accessKeyJson);
-    try {
-      this.keyConfig.write();
-    } catch (error) {
-      throw new Error(`Failed to save config: ${error}`);
-    }
+    this.accessKeys.push(accessKey);
+    this.saveAccessKeys();
     await this.updateServer();
-    return makeAccessKey(this.proxyHostname, accessKeyJson);
+    return accessKey;
   }
 
   removeAccessKey(id: AccessKeyId): boolean {
-    for (let ai = 0; ai < this.keyConfig.data().accessKeys.length; ai++) {
-      const accessKey = this.keyConfig.data().accessKeys[ai];
+    for (let ai = 0; ai < this.accessKeys.length; ai++) {
+      const accessKey = this.accessKeys[ai];
       if (accessKey.id === id) {
-        this.portProvider.freePort(accessKey.port);
-        this.keyConfig.data().accessKeys.splice(ai, 1);
-        this.keyConfig.write();
+        this.portProvider.freePort(accessKey.proxyParams.portNumber);
+        this.accessKeys.splice(ai, 1);
+        this.saveAccessKeys();
         this.updateServer();
         return true;
       }
     }
     return false;
   }
 
-  listAccessKeys(): IterableIterator<AccessKey> {
-    return this.keyConfig.data().accessKeys.map(
-        accessKeyJson => makeAccessKey(this.proxyHostname, accessKeyJson))[Symbol.iterator]();
+  listAccessKeys(): AccessKey[] {
+    // Deep copy the access key array to prevent returning references to keys or their properties.
```

Comment:
Why is it bad to keep references to the keys?

Reply:
If we return references to the keys, a caller could bypass the repository's methods to modify them.

Reply:
Ugh. I feel like we should change the API so that the mutations are methods on the AccessKey object. But that's a change for another PR.